### PR TITLE
Federation experiment

### PR DESCRIPTION
**Pull request by gtklocker**

### CODE REVIEW

File: experiment/contracts/bitcoin-spv/CheckBitcoinSigsDelegate.sol
Comment (dionyziz) [inline experiment/contracts/bitcoin-spv/CheckBitcoinSigsDelegate.sol]: Is it possible that we can somehow import these smart contracts without including them in our codebase?

File: experiment/test/utils.js
Comment (dionyziz) [inline experiment/test/utils.js]: eol

File: experiment/test/tx-fixtures/valid-burn.json
Comment (dionyziz) [inline experiment/test/tx-fixtures/valid-burn.json]: eol

File: experiment/test/tx-fixtures/1-input-2-outputs-first-p2pkh.json
Comment (dionyziz) [inline experiment/test/tx-fixtures/1-input-2-outputs-first-p2pkh.json:1]: Is there a better way to make these fixtures more readable? Perhaps vout can be generated at runtime using bcoin.js.

File: experiment/test/Verifier.test.js
Comment (dionyziz) [inline experiment/test/Verifier.test.js:133]: check that invalidity returns false also

File: experiment/test/Crosschain.test.js
Comment (dionyziz) [inline experiment/test/Crosschain.test.js]: with a proof

Comment (dionyziz) [reply experiment/test/Crosschain.test.js]: unnecessary "."

File: experiment/test/BurnedBTC.test.js
Comment (dionyziz) [inline experiment/test/BurnedBTC.test.js]: start

File: experiment/scripts/tx-parse.py
Comment (dionyziz) [inline experiment/scripts/tx-parse.py]: eol

File: experiment/scripts/tx-parse.py
Comment (dionyziz) [inline experiment/scripts/tx-parse.py:96]: I feel this would be a good fit for an OOP implementation of a "Transaction" class with a "deserialize" method.

File: experiment/scripts/tx-parse.py
Comment (dionyziz) [inline experiment/scripts/tx-parse.py]: Avoid else after return, especially when returning early on error condition such as here.

File: experiment/scripts/tx-parse.py
Comment (dionyziz) [inline experiment/scripts/tx-parse.py:15]: Is this standard terminology? I would call it "last". But I dislike that this updates the state of the class. Shouldn't this be done just by pop?

File: experiment/scripts/generate-burn-proof.py
Comment (dionyziz) [inline experiment/scripts/generate-burn-proof.py]: nit double spacing

File: experiment/scripts/generate-burn-proof.py
Comment (dionyziz) [inline experiment/scripts/generate-burn-proof.py]: eol

File: experiment/scripts/gen-burn-address.py
Comment (dionyziz) [inline experiment/scripts/gen-burn-address.py]: eol

File: experiment/scripts/gen-burn-address.py
Comment (dionyziz) [inline experiment/scripts/gen-burn-address.py]: Let's get rid of this \x04 now.

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: Use `~` before `\cite` to avoid a breaking space.

Comment (dionyziz) [inline empirical-results.tex]: When you say "we provide" it is implied that this is provided within the paper. This is not the case.

Comment (dionyziz) [reply empirical-results.tex]: Write open source without a hyphen

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: This paragraph can be moved to a footnote. Also let's anonymize this paper by introducing an `\ifanonymous` macro at this point to avoid name leaks in the FC submission.

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: This seems like it could have an impact on gas costs and overall appears like a serious limitation not representative of the federation approach. A reviewer would find this peculiar. Why don't we have a federation of, say, 3?

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: Shorten this: Using prices at the time of writing, we obtain a median gas price of $6.9$ gwi and the price of $1$ ETH = $\$170.07$, giving a gas price of $\$1.173483 * 10^{-6}$.

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: What is this address?

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex:3]: ?

File: experiment/contracts/BurnVerifier.sol
Comment (dionyziz) [inline experiment/contracts/BurnVerifier.sol]: Drop the 04

File: experiment/contracts/BurnedBTC.sol
Comment (dionyziz) [inline experiment/contracts/BurnedBTC.sol]: eol

File: experiment/contracts/BurnedBTC.sol
Comment (dionyziz) [inline experiment/contracts/BurnedBTC.sol:8]: Do we want this to extend ERC20 also?

File: experiment/contracts/CheckpointRepo.sol
Comment (dionyziz) [inline experiment/contracts/CheckpointRepo.sol]: eol

File: experiment/contracts/CheckpointRepo.sol
Comment (dionyziz) [inline experiment/contracts/CheckpointRepo.sol:3]: It feels easy to add a federation of 3 and this would make the claims in our paper seem much more plausible without too much effort. Otherwise we give the reviewer an easy way out to reject.

File: experiment/contracts/Crosschain.sol
Comment (dionyziz) [inline experiment/contracts/Crosschain.sol]: eol

File: experiment/contracts/Verifier.sol
Comment (dionyziz) [inline experiment/contracts/Verifier.sol]: eol

File: experiment/contracts/Crosschain.sol
Comment (dionyziz) [inline experiment/contracts/Crosschain.sol:11]: We don't support multiple burn outputs per tx?

File: experiment/contracts/Crosschain.sol
Comment (dionyziz) [inline experiment/contracts/Crosschain.sol]: Let's call events `e` following the paper.

File: experiment/contracts/Verifier.sol
Comment (dionyziz) [inline experiment/contracts/Verifier.sol]: I don't understand who uses this and why it's needed.

File: experiment/contracts/Verifier.sol
Comment (dionyziz) [inline experiment/contracts/Verifier.sol]: `bytes memory actualValue = output.extractValue()`

**Review by dionyziz — CHANGES_REQUESTED** [review]
Let's walk over this code together so that I fully understand it.

File: experiment/contracts/Crosschain.sol
Comment (gtklocker) [reply experiment/contracts/Crosschain.sol:11]: we do

File: experiment/contracts/Verifier.sol
Comment (gtklocker) [reply experiment/contracts/Verifier.sol]: rm'd

File: experiment/scripts/tx-parse.py
Comment (gtklocker) [reply experiment/scripts/tx-parse.py:15]: Excursion is a term from emacs but not used in the same way here. An excursion here is the concatenation of all pops that have happened since the last excursion was executed. This allows us to do transaction parsing while keeping the parsed parts easily as you can see in the code after it. However this code is not part of the smart contract and is only a helper script I used during development so I'm not too concerned.

File: experiment/contracts/BurnedBTC.sol
Comment (gtklocker) [reply experiment/contracts/BurnedBTC.sol:8]: Not for our experiments, but for actual deployments it could.

File: empirical-results.tex
Comment (gtklocker) [reply empirical-results.tex]: explained IRL, would you prefer this removed?

Comment (gtklocker) [reply empirical-results.tex]: could you remind me what the "Accessed at" footnote should look like exactly? Or should it be a citation?

File: empirical-results.tex
Comment (gtklocker) [reply empirical-results.tex]: What would you like to be shown on the anon version?

File: empirical-results.tex
Comment (gtklocker) [reply empirical-results.tex:3]: Some refs are included depend on another branch. I used this macro to be able to easily disable them until the other branch is merged

Comment (gtklocker) [reply empirical-results.tex:3]: Instead of creating a dependency or rebasing on top of the other

File: experiment/scripts/tx-parse.py
Comment (gtklocker) [reply experiment/scripts/tx-parse.py:96]: I don't necessarily agree, but this helper script is not worth the effort imho anyway

File: experiment/test/tx-fixtures/1-input-2-outputs-first-p2pkh.json
Comment (gtklocker) [reply experiment/test/tx-fixtures/1-input-2-outputs-first-p2pkh.json:1]: perhaps using bcoin or some other library would help, but not worth the effort at this point imo

File: empirical-results.tex
Comment (gtklocker) [reply empirical-results.tex]: What should I write instead of "we provide"?

File: experiment/contracts/bitcoin-spv/CheckBitcoinSigsDelegate.sol
Comment (gtklocker) [reply experiment/contracts/bitcoin-spv/CheckBitcoinSigsDelegate.sol]: done

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: Refs should name the thing you ref. `Algorithm~\ref{alg.burn-verifier}`.

Comment (dionyziz) [reply empirical-results.tex]: Should we use the same names in the implemented smart contracts and the pseudocode smart contracts? Why do we have two names? Are `BurnedBTC` and `burn-verifier` doing the same thing?

Comment (dionyziz) [reply empirical-results.tex]: "where the leaf corresponding to each block has the value of the block's transaction id merkle tree root."

too verbose, split into multiple sentences, say,

"where each leaf corresponds to a block. The leaf stores the merkle tree root of the merkle tree which collects the transaction id of the transactions included in the block in question."

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: Did we write the BurnVerifier library? If so, why do we say it "provides"? Start by say "We put forth the library \textsf{BurnVerifier} which provides..."

File: empirical-results.tex
Comment (dionyziz) [reply empirical-results.tex]: Yes. Say "The link to our open source software implementation, which uses the MIT license, has been redacted for anonymity. The implementation is production-ready and fully tested with 100\% code coverage."

**Review by dionyziz — CHANGES_REQUESTED** [review]

**Comment by gtklocker** [comment]
The only thing I haven't addressed is the naming weirdness (different names for latex algos and implementation contracts) but I think it's more productive to open an issue for it than have it block this PR.

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: Capitalize Merkle Tree

Comment (dionyziz) [reply empirical-results.tex]: Explain what "block connection" means

File: empirical-results.tex
Comment (dionyziz) [inline empirical-results.tex]: Unless we show these contracts as pseudocode within the paper, there's little point in citing their names. The reader cannot discern what they do. We should try to give a higher-level picture here.

**Review by dionyziz — APPROVED** [review]

File: empirical-results.tex
Comment (gtklocker) [reply empirical-results.tex]: block connection is defined in the previous section

**Pull request merged**
